src/main/java/com/google/step/YOUR_PROJECT_NAME_HERE/external/W3SchoolClient.java

```diff
@@ -0,0 +1,43 @@
+package com.google.step.YOUR_PROJECT_NAME_HERE.external;
+
+import com.google.step.YOUR_PROJECT_NAME_HERE.data.Card;
+import java.io.IOException;
+import org.jsoup.Jsoup;
+import org.jsoup.nodes.Document;
+import org.jsoup.select.Elements;
+
+public final class W3SchoolClient {
+  private static final String TITLE_ID = "h1";
+  private static final String DESC_ID = "p";
+  private static final String SNIPPET_ID = "w3-example";
+  private static final String CODE_ID = "w3-code";
+
+  public Card search(String w3Link) throws IOException {
+
+    Document doc = Jsoup.connect(w3Link).get();
+    Elements titles = doc.getElementsByTag(TITLE_ID);
+    Elements descriptions = doc.getElementsByTag(DESC_ID);
+    Elements snippets = doc.getElementsByClass(SNIPPET_ID);
+
+    String title = "";
+    String description = "";
+    String code = "";
+
+    if (titles != null
+        && !titles.isEmpty()
+        && descriptions != null
+        && !descriptions.isEmpty()
+        && snippets != null
+        && !snippets.isEmpty()) {
+      title = titles.first().text();
+      description = descriptions.first().text();
+      code = snippets.first().getElementsByClass(CODE_ID).text();
+    }
+
+    if (!title.isEmpty() && !description.isEmpty() && !code.isEmpty()) {
+      Card w3Card = new Card(title, code, w3Link, description);
+      return w3Card;
```

Comment:
what if you cannot retrieve some of the field? i think the returned card can have some empty field.

Reply:
I don't see any value in returning / displaying half empty cards. I think we already identified 4 essential properties of a Card and I think returning only title/link or returning only code with no title would be more confusing and seemingly incomplete than helpful

Reply:
i'm thinking of some page that might not have code in it?

Reply:
*resolved via discussion*